torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp

```diff
@@ -0,0 +1,592 @@
+        status == NVJPEG_STATUS_SUCCESS,
+        "Failed to attach device buffer: ",
+        status);
+    int buffer_index = 0;
+    // UNCHANGED behaves weird, so we use RGB instead
+    status = nvjpegDecodeParamsSetOutputFormat(
+        nvjpeg_decode_params,
+        output_format == NVJPEG_OUTPUT_UNCHANGED ? NVJPEG_OUTPUT_RGB
+                                                 : output_format);
+    TORCH_CHECK(
+        status == NVJPEG_STATUS_SUCCESS,
+        "Failed to set output format: ",
+        status);
+    for (std::vector<at::Tensor>::size_type i = 0; i < sw_input_buffer.size();
+         ++i) {
+      status = nvjpegJpegStreamParse(
+          nvjpeg_handle,
+          sw_input_buffer[i],
+          sw_input_buffer_size[i],
+          0,
+          0,
+          jpeg_streams[buffer_index]);
+      TORCH_CHECK(
+          status == NVJPEG_STATUS_SUCCESS,
+          "Failed to parse jpeg stream: ",
+          status);
+
+      status = nvjpegStateAttachPinnedBuffer(
+          nvjpeg_decoupled_state, pinned_buffers[buffer_index]);
+      TORCH_CHECK(
+          status == NVJPEG_STATUS_SUCCESS,
+          "Failed to attach pinned buffer: ",
+          status);
+
+      status = nvjpegDecodeJpegHost(
+          nvjpeg_handle,
+          nvjpeg_decoder,
+          nvjpeg_decoupled_state,
+          nvjpeg_decode_params,
+          jpeg_streams[buffer_index]);
+      TORCH_CHECK(
+          status == NVJPEG_STATUS_SUCCESS,
+          "Failed to decode jpeg stream: ",
+          status);
+
+      cudaStatus = cudaStreamSynchronize(stream);
+      TORCH_CHECK(
+          cudaStatus == cudaSuccess,
+          "Failed to synchronize CUDA stream: ",
+          cudaStatus);
+
+      status = nvjpegDecodeJpegTransferToDevice(
+          nvjpeg_handle,
+          nvjpeg_decoder,
+          nvjpeg_decoupled_state,
+          jpeg_streams[buffer_index],
+          stream);
+      TORCH_CHECK(
+          status == NVJPEG_STATUS_SUCCESS,
+          "Failed to transfer jpeg to device: ",
+          status);
+
+      buffer_index = 1 - buffer_index; // switch pinned buffer in pipeline mode
+                                       // to avoid an extra sync
+
+      status = nvjpegDecodeJpegDevice(
+          nvjpeg_handle,
+          nvjpeg_decoder,
+          nvjpeg_decoupled_state,
+          &sw_output_buffer[i],
+          stream);
+      TORCH_CHECK(
+          status == NVJPEG_STATUS_SUCCESS,
+          "Failed to decode jpeg stream: ",
+          status);
+    }
+  }
+
+  cudaStatus = cudaStreamSynchronize(stream);
+  TORCH_CHECK(
+      cudaStatus == cudaSuccess,
+      "Failed to synchronize CUDA stream: ",
+      cudaStatus);
+
+  // prune extraneous channels from single channel images
+  if (output_format == NVJPEG_OUTPUT_UNCHANGED)
+    for (std::vector<at::Tensor>::size_type i = 0; i < output_tensors.size();
+         ++i)
+      if (channels[i] == 1)
+        output_tensors[i] = output_tensors[i][0].unsqueeze(0).clone();
```

Comment:
Out of curiosity -- why is is clone needed here? clone could be expensive hence my question

Reply:
To be quite honest, I don't remember and at this point I'll probably just go ahead and land this and we can send a follow up PR if needed. In any event, this would only apply to grayscale images